File: src/lessons/welcome/loopfor/LoopCourseForestEntity.py
import javax.swing.JOptionPane;

colors = [Color(0,155,0), Color(50,155,0), Color(100,155,0), Color(140,155,0),
          Color(160,155,0), Color(180,155,0), Color(200,155,0), Color(210,155,0), Color.red]

def forward(i=1):
    if i>1:
      errorMsg("Sorry Dave, I cannot let you use forward with an argument in this exercise. Use a loop instead.")
    else:
        if not haveSeenError():
            entity.forward()
        c = getGroundColor()
        if c == Color.blue:
            if not haveSeenError():
                javax.swing.JOptionPane.showMessageDialog(None, "You fall into water.", "Test failed", javax.swing.JOptionPane.ERROR_MESSAGE);
            seenError();
        else:
          for i in range(len(colors)-1):
              if colors[i] == c:
                  if i == len(colors)-1:
                      c = colors[i]
                  else:
                      c = colors[i+1]
                  break
          setBrushColor(c)    
          brushDown()
          brushUp()
def backward(i=1):
    if i>1:
      errorMsg("Sorry Dave, I cannot let you use backward with an argument in this exercise. Use a loop instead.")
    else:
      errorMsg("Sorry Dave, you cannot run backward that way. Exercising is hard enough -- please don't overplay.")
# BEGIN SOLUTION
for i in range(7):
    for side in range(4):
        for step in range(4):
            forward()
        left();
        for step in range(2):
            forward()
        right();
        for step in range(4):
            forward()
        right()
        forward()
        forward()
        left()
        for step in range(4):
            forward()
        left()
# END SOLUTION

File: src/lessons/welcome/loopfor/LoopCourseForestEntity.c
#include "../../../../lib/resources/langages/c/include/RemoteBuggle.h"

char* colors[] = {"0/155/0",   "50/155/0",  "100/155/0", "140/155/0", "160/155/0",
                  "180/155/0", "200/155/0", "210/155/0", "255/0/0"};

void forward_observer(int steps)
{
  if (steps > 1) {
    printf("Sorry Dave, I cannot let you use forward with an argument in this exercise. Use a loop instead.");
    exit(1);
  }

  const int color_length = sizeof(colors) / sizeof(colors[0]);
  char* old              = getGroundColorName();

  if (strcmp(old, "0/0/255") == 0) { /* fallen into water */
    free(old);
    return;
  }

  char* new = old;
  for (int i = 0; i < color_length - 1; i++)
    if (strcmp(colors[i], old) == 0) {
      if (i == color_length - 1)
        new = colors[i];
      else
        new = colors[i + 1];
      break;
    }
  setBrushColorName(new);
  brushDown();
  brushUp();
  free(old);
}

void backward_observer(int steps)
{
  printf("Sorry Dave, I cannot let you use back with an argument in this exercise. Exercising is hard enough -- please "
         "don't overplay.");
  exit(1);
}

void run()
{
//   set_post_forward(&forward_observer);
//   set_pre_backward(&backward_observer);
  /* BEGIN TEMPLATE */
  /* BEGIN SOLUTION */
  for (int i = 0; i < 7; i++)
    for (int side = 0; side < 4; side++) {
      for (int step = 0; step < 4; step++)
        forward(1);
      left();
      for (int step = 0; step < 2; step++)
        forward(1);
      right();
      for (int step = 0; step < 4; step++)
        forward(1);
      right();
      forward(1);
      forward(1);
      left();
      for (int step = 0; step < 4; step++)
        forward(1);
      left();
    }
  /* END SOLUTION */
  /* END TEMPLATE */
}
